File: problem2/ExpressionAddOperators.java
//Time Complexity : O(n*(4^n)), n -> Size of string
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
package problem2;

import java.util.ArrayList;
import java.util.List;

public class ExpressionAddOperators {
	List<String> ans;

	public List<String> addOperators(String num, int target) {
		ans = new ArrayList<>();
		if (num == null || num.length() == 0) {
			return ans;
		}

		helper(num, target, 0, 0, 0, new StringBuilder());
		return ans;
	}

	public void helper(String num, int target, int index, long calc, long tail, StringBuilder expression) {
		// Base
		if (index == num.length()) {
			if (calc == target) {
				ans.add(expression.toString());
			}
			return;
		}
		// Logic
		for (int i = index; i < num.length(); i++) {
			if (num.charAt(index) == '0' && i != index) {
				return;
			}

			long curr = Long.parseLong(num.substring(index, i + 1));
			int len = expression.length();
			if (index == 0) {
				expression.append(curr);
				helper(num, target, i + 1, curr, curr, expression);
				expression.setLength(len);
			} else {
				// Addition
				expression.append("+");
				expression.append(curr);
				helper(num, target, i + 1, calc + curr, curr, expression);
				expression.setLength(len);
				// Subtraction
				expression.append("-");
				expression.append(curr);
				helper(num, target, i + 1, calc - curr, -curr, expression);
				expression.setLength(len);
				// Multiplication
				expression.append("*");
				expression.append(curr);
				helper(num, target, i + 1, calc - tail + (tail * curr), tail * curr, expression);
				expression.setLength(len);
			}
		}
	}

	public static void main(String[] args) {
		ExpressionAddOperators obj = new ExpressionAddOperators();
		String num = "123";
		int target = 6;
		List<String> ans = obj.addOperators(num, target);

		System.out.println("Set of expressions that evaluate to the target \'" + target + "\'");
		for (String str : ans) {
			System.out.println(str);
		}
		System.out.println();
	}

}

File: ExpressionAddOperators.cs
// Time Complexity : O(N!)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

    //backtracking
    IList<string> result;
    public IList<string> AddOperators(string num, int target) {
        result = new List<string>();
        if(num == null || num.Length == 0)
            return result;
        
        backtracking(num, target, 0, 0, 0, "");

        return result;
    } 
    private void backtracking(string num, int target, StringBuilder sb, long calc, long tail, int index)
    {
        //base
        if(index == num.Length)
        {
            if(target == calc)
            {
                result.Add(sb.ToString());
            }
            return;
        }               
            
        //logic
        for(int i = index; i < num.Length; i++)
        {            
            //this line is to avoid scenarios, where first element is 0 and we are appending non zero elemen to it
            //05, 0 is index and i is at 5, when we convert to log, it becomes 5, which is wrong
            //so wwe avoid sitation , where num index is 0 and i is not at index
            if(index != i && num[index] == '0') continue;
            int len = sb.Length;
            long curr = (long)Convert.ToDouble(num.Substring(index, i+1-index));            
            if(index == 0)
            {
                sb.Append(curr);
                backtracking(num, target, sb, curr, curr, i+1);
                sb.Length = len;
            }
            else
            {
                //+ case
                //action
                sb.Append("+");
                sb.Append(curr);
                //recurse
                backtracking(num, target, sb, calc + curr, curr, i+1);
                //backtrack
                sb.Length = len;
                
                //- case
                //action
                sb.Append("-");
                sb.Append(curr);
                //recurse
                backtracking(num, target, sb, calc - curr, -curr, i+1);
                //backtrack
                sb.Length = len;
                
                //* case
                //action
                sb.Append("*");
                sb.Append(curr);
                //recurse
                backtracking(num, target, sb, (calc - tail) + (tail * curr), tail * curr, i+1);
                //backtrack
                sb.Length = len;
            }
        }
        
    }
    

    //recursion only
    public IList<string> AddOperators(string num, int target) {
        result = new List<string>();
        if(num == null || num.Length == 0)
            return result;
        
        helper(num, target, 0, 0, 0, "");
        backtracking(num, target, 0, 0, 0, "");

        return result;
        
    } 
    private void helper(string num, int target, long calc, long tail, int index, string path)
    {
        //base
        if(index == num.Length)
        {
            if(target == calc)
            {
                result.Add(path);
            }
            return;
        }               
            
        //logic
        for(int i = index; i < num.Length; i++)
        {            
            //this line is to avoid scenarios, where first element is 0 and we are appending non zero elemen to it
            //05, 0 is index and i is at 5, when we convert to log, it becomes 5, which is wrong
            //so wwe avoid sitation , where num index is 0 and i is not at index
            if(index != i && num[index] == '0') continue;
            long curr = (long)Convert.ToDouble(num.Substring(index, i+1-index));            
            if(index == 0)
            {
                helper(num, target, curr, curr, i+1, path+curr);
            }
            else
            {
                //+ case
                helper(num, target, calc + curr, curr, i+1, path + "+" + curr);
                //- case
                helper(num, target, calc - curr, -curr, i+1, path + "-" + curr);
                //* case
                helper(num, target, (calc - tail) + (tail * curr), tail * curr, i+1, path + "*" + curr);
            }
        }
        
    }
    
    private void backtracking(string num, int target, long calc, long tail, int index, string path)
    {
        //base
        if(index == num.Length)
        {
            if(target == calc)
            {
                result.Add(path);
            }
            return;
        }               
            
        //logic
        for(int i = index; i < num.Length; i++)
        {            
            //this line is to avoid scenarios, where first element is 0 and we are appending non zero elemen to it
            //05, 0 is index and i is at 5, when we convert to log, it becomes 5, which is wrong
            //so wwe avoid sitation , where num index is 0 and i is not at index
            if(index != i && num[index] == '0') continue;
            long curr = (long)Convert.ToDouble(num.Substring(index, i+1-index));            
            if(index == 0)
            {
                helper(num, target, curr, curr, i+1, path+curr);
            }
            else
            {
                //+ case
                helper(num, target, calc + curr, curr, i+1, path + "+" + curr);
                //- case
                helper(num, target, calc - curr, -curr, i+1, path + "-" + curr);
                //* case
                helper(num, target, (calc - tail) + (tail * curr), tail * curr, i+1, path + "*" + curr);
            }
        }
        
    }